src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestLocator.kt

```diff
@@ -0,0 +1,45 @@
+package org.rust.cargo.runconfig.test
+
+import com.intellij.execution.Location
+import com.intellij.execution.PsiLocation
+import com.intellij.execution.testframework.sm.runner.SMTestLocator
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.GlobalSearchScope
+import org.rust.lang.core.psi.RsFunction
+import org.rust.lang.core.resolve.ResolveEngine
+
+/**
+ * Url format:
+ *  - test functions: `cargo:test://{test function fq name}`
+ *
+ * **It is preferable to use `get...Url` methods instead of manually constructing urls.**
+ */
+object CargoTestLocator : SMTestLocator {
+    const val TEST_PROTOCOL = "cargo:test"
+
+    fun getLocation(url: String, project: Project, scope: GlobalSearchScope): List<Location<PsiElement>> =
+        getLocation(
+            requireNotNull(VirtualFileManager.extractProtocol(url)) { "couldn't extract protocol" },
+            VirtualFileManager.extractPath(url),
+            project,
+            scope)
+
+    override fun getLocation(
+        protocol: String,
+        path: String,
+        project: Project,
+        scope: GlobalSearchScope
+    ): List<Location<PsiElement>> {
+        if (protocol == TEST_PROTOCOL) {
+            val resolveResult = ResolveEngine.resolve(path, project) ?: return emptyList()
+            return listOf(PsiLocation.fromPsiElement<PsiElement>(project, resolveResult.element))
+        }
```

Comment:
Basically there are two ways to search for a thing in Rust: given a crate root and a path, resolve the path inside the crate. This is what Resolve Engine does.

Another option is to search for the last identifier in the index (we can't store the whole path in index because it depends on more than one file. Compare this with Kotlin, where you always have a handy `package` clause at the beginning of the file).

`SearchScopes` are used in the second case in the query to the index to do `VirtualFile` based filtering. 

I think either approach would work in Rust, but we need to know which Cargo package/target are we talking about. This info should be available to `CargoTestConsoleProperties` because it knows the cargo command line it will execute.